### PR TITLE
Make grpc compile on win64bit

### PR DESCRIPTION
**Pull request by jtattermusch**
"long" type is 32bit on Win x64, which brings some trouble.

Please note the new public function gpr_int64toa.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Travis failure unrelated: #4295

**Comment by jtattermusch** [comment]
Interop test failure seems unrelated, @murgatroid99  is working on fixing the problem.

**Comment by jtattermusch** [comment]
friendly ping.

File: include/grpc++/impl/rpc_service_method.h [hunk 1]
```diff
@@ -251,7 +252,11 @@ class RpcService {
   void AddMethod(RpcServiceMethod* method) { methods_.emplace_back(method); }
 
   RpcServiceMethod* GetMethod(int i) { return methods_[i].get(); }
-  int GetMethodCount() const { return methods_.size(); }
+  int GetMethodCount() const {
+    // On win x64, int is only 32bit
+    GPR_ASSERT(methods_.size() <= INT_MAX);
```
Comment (nicolasnoble) [inline include/grpc++/impl/rpc_service_method.h]: *Cough* at least this isn't 16 bits *COUGH* *COUGH*

File: src/core/iomgr/tcp_windows.c [hunk 2]
```diff
@@ -281,19 +281,20 @@ static void win_write(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
...
+  GPR_ASSERT(tcp->write_slices->count <= UINT32_MAX);
   if (tcp->write_slices->count > GPR_ARRAY_SIZE(local_buffers)) {
     buffers = (WSABUF *)gpr_malloc(sizeof(WSABUF) * tcp->write_slices->count);
     allocated = buffers;
   }
 
   for (i = 0; i < tcp->write_slices->count; i++) {
-    buffers[i].len = GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
+    GPR_ASSERT(GPR_SLICE_LENGTH(tcp->write_slices->slices[i]) <= UINT32_MAX);
+    buffers[i].len = (ULONG) GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
```
Comment (nicolasnoble) [inline src/core/iomgr/tcp_windows.c]: size_t len = GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
GPR_ASSERT(len <= ULONG_MAX);
buffers[i].len = len;

Comment (jtattermusch) [inline src/core/iomgr/tcp_windows.c]: Done.

**Comment by jtattermusch** [comment]
h2_proxy_simple_request_test
h2_proxy_simple_request_nosec_test  seem to be consistently failing. Could that be because of this PR?

**Comment by jtattermusch** [comment]
Looks like h2_proxy_simple_request_test and h2_proxy_simple_request_nosec_test is also failing on master here: https://grpc-testing.appspot.com/job/gRPC_master/13556/#showFailuresLink

**Comment by jtattermusch** [comment]
I think I've addressed all the comments. PTAL.

File: src/core/iomgr/tcp_windows.c [hunk 3]
```diff
@@ -197,7 +197,7 @@ static void win_read(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
 
   tcp->read_slice = gpr_slice_malloc(8192);
 
-  buffer.len = GPR_SLICE_LENGTH(tcp->read_slice);
+  buffer.len = (ULONG)GPR_SLICE_LENGTH(tcp->read_slice);  // we know slice size fits in 32bit.
```
Comment (ctiller) [inline src/core/iomgr/tcp_windows.c]: I'm not sure this is true: it'd be worth an assert at least

File: src/core/support/string.c [hunk 4]
```diff
@@ -173,6 +173,27 @@ int gpr_ltoa(long value, char *string) {
   return i;
 }
 
+int gpr_int64toa(gpr_int64 value, char *string) {
```
Comment (ctiller) [inline src/core/support/string.c]: I don't see any tests for this function

File: src/core/iomgr/tcp_windows.c (hunk 3 above)
Comment (jtattermusch) [inline src/core/iomgr/tcp_windows.c]: we've allocated read_slice of size 8192 on the previous line. should I still add the assert?

File: src/core/support/string.c (hunk 4 above)
Comment (jtattermusch) [inline src/core/support/string.c]: Yes, I can add some once I get a thumbs up for adding this PUBLIC function.

**Comment by jtattermusch** [comment]
Turns out there were no tests for gpr_ltoa, so I added tests for both gpr_ltoa and gpr_int64toa.
In the process of doing that, it turned out that handling for LONG_MIN and INT64_MIN was broken, so I fixed that. PTAL.

File: src/core/iomgr/tcp_windows.c (hunk 3 above)
Comment (ctiller) [inline src/core/iomgr/tcp_windows.c]: oh ffs I can't read

File: test/core/support/string_test.c [hunk 5]
```diff
@@ -286,6 +286,53 @@ static void test_strsplit(void) {
   gpr_free(parts);
 }
 
+test_ltoa() {
```
Comment (ctiller) [inline test/core/support/string_test.c]: void test_ltoa(void) {

File: test/core/support/string_test.c [hunk 6]
```diff
@@ -286,6 +286,53 @@ static void test_strsplit(void) {
...
+  GPR_ASSERT(0 == strcmp("-12345", buf));
+
+  /* large negative - we don't know the size of long in advance */
+  GPR_ASSERT(gpr_asprintf(&str, "%lld", (long long)LONG_MIN));
+  GPR_ASSERT(strlen(str) == gpr_ltoa(LONG_MIN, buf));
+  GPR_ASSERT(0 == strcmp(str, buf));
+  gpr_free(str);
+}
+
+test_int64toa() {
```
Comment (ctiller) [inline test/core/support/string_test.c]: void test_int64toa(void) {

File: test/core/support/string_test.c (hunk 5 above)
Comment (jtattermusch) [inline test/core/support/string_test.c]: ah, sorry. made it static void like all the other test methods.

File: test/core/support/string_test.c (hunk 6 above)
Comment (jtattermusch) [inline test/core/support/string_test.c]: done

**Comment by ctiller** [comment]
Now Linux fails:

```c
src/core/support/string.c: In function 'gpr_ltoa':
src/core/support/string.c:167:12: error: conversion to 'long unsigned int' from 'long int' may change the sign of the result [-Werror=sign-conversion]
     uval = -value;
            ^
src/core/support/string.c:169:12: error: conversion to 'long unsigned int' from 'long int' may change the sign of the result [-Werror=sign-conversion]
     uval = value;
            ^
src/core/support/string.c: In function 'gpr_int64toa':
src/core/support/string.c:193:12: error: conversion to 'gpr_uint64' from 'gpr_int64' may change the sign of the result [-Werror=sign-conversion]
     uval = -value;
            ^
src/core/support/string.c:195:12: error: conversion to 'gpr_uint64' from 'gpr_int64' may change the sign of the result [-Werror=sign-conversion]
     uval = value;
```

**Comment by jtattermusch** [comment]
Alright, I did some fix but it looks like this still relies implementation-specific. I will try rewrite the implementation to be 100% compliant with the C spec.

**Comment by jtattermusch** [comment]
I got rid of the negation from ltoa and int64toa code, hoping that now everything should be compliant with the spec.

**Comment by jtattermusch** [comment]
Weird error building openSSL, I don't see how this could be related.

**Comment by jtattermusch** [comment]
Jenkins is sufficiently green.

**Pull request merged**